demos/cms/cms_ver.c

```diff
@@ -12,6 +12,53 @@
 #include <openssl/cms.h>
 #include <openssl/err.h>
 
+/*
+ * print any signingTime attributes.
+ * signingTime is when each party purportedly signed the message.
+ */
+static void print_signingTime(CMS_ContentInfo *cms)
+{
+    STACK_OF(CMS_SignerInfo) *sis;
+    CMS_SignerInfo *si;
+    X509_ATTRIBUTE *attr;
+    ASN1_TYPE *t;
+    ASN1_UTCTIME *utctime;
+    ASN1_GENERALIZEDTIME *gtime;
+    BIO *b;
+    int i, loc;
+
+    b = BIO_new_fp(stdout, BIO_NOCLOSE);
+    sis = CMS_get0_SignerInfos(cms);
+    for (i = 0; i < sk_CMS_SignerInfo_num(sis); i++) {
+        si = sk_CMS_SignerInfo_value(sis, i);
+        loc = CMS_signed_get_attr_by_NID(si, NID_pkcs9_signingTime, -1);
+        attr = CMS_signed_get_attr(si, loc);
+        if (X509_ATTRIBUTE_count(attr) > 1) {
+            fprintf(stderr, "error: at most one signingTime attribute is permitted\n");
+            continue;
+        }
+        t = X509_ATTRIBUTE_get0_type(attr, 0);
+        if (t == NULL)
+            continue;
+        switch (t->type) {
+        case V_ASN1_UTCTIME:
+            utctime = t->value.utctime;
+            ASN1_UTCTIME_print(b, utctime);
+            break;
+        case V_ASN1_GENERALIZEDTIME:
+            gtime = t->value.generalizedtime;
+            ASN1_GENERALIZEDTIME_print(b, utctime);
+            break;
+        default:
+            fprintf(stderr, "unrecognized signingTime type\n");
+            break;
+        }
+        printf(": signingTime from SignerInfo %i\n", i);
```

Comment:
BIO_printf() ?

Reply:
good idea.

`BIO_printf(b, ...)` would better match the calls to `ASN1_UTCTIME_print(b, utctime)` and `ASN1_GENERALIZEDTIME_print(b, gtime)`.

fixed.